### ISSUE TITLE
Comprehensive Baseline Comparison for NeurIPS Submission

### ISSUE THREAD

**Issue by AbdullahKaratas** [AbdullahKaratas/neural-spectral-gp#4]
## 📋 Overview

We need a systematic benchmark suite comparing F-SDN against established methods to demonstrate our contribution for the NeurIPS 2026 paper. The current implementation has stable code (99% Silverman error) and all critical bugs fixed, making this the right time to build comprehensive baselines.

## 🎯 Motivation

**Current Status:**
- ✅ F-SDN implementation complete and validated (99% error on Silverman benchmark)
- ✅ All critical bugs fixed (Bug #4: factor-4 scaling, Bug #5: missing sin-basis)
- ✅ Method consistency verified (Low-Rank, Deterministic, MC all agree)
- ✅ Paper skeleton complete with LaTeX compiling cleanly

**What's Missing:**
- ❌ No systematic comparison against state-of-the-art methods
- ❌ Limited benchmark coverage (only 2 synthetic tests)
- ❌ No real-world dataset validation
- ❌ Missing performance metrics (time, memory, parameters)
- ❌ No LaTeX table for experimental results section

**Why This Matters:**
The experimental section is the cornerstone of our NeurIPS submission. Without comprehensive baselines, we cannot:
1. Demonstrate competitive performance against existing methods
2. Justify our architectural choices empirically
3. Show where F-SDN excels vs. struggles
4. Meet reviewer expectations for thoroughness

---

### 1. Methods to Compare

**Primary Baselines:**
- **F-SDN** (ours) - Factorized Spectral Density Network with guaranteed PSD
- **Standard GP** - Vanilla Gaussian Process with RBF/Matérn kernels (GPyTorch)
- **RFF** - Random Fourier Features (Rahimi & Recht, 2007)
- **Remes et al.** - Spectral Mixture approach (official implementation already integrated)

**Optional (if time permits):**
- **Sparse GP** - Inducing point methods (SVGP)
- **Deep Kernel Learning** - Neural network + GP (Wilson et al., 2016)

### 2. Benchmark Datasets

**Synthetic (Controlled):**
- ✅ Silverman Non-Stationary (already done: 99% error)
- ✅ Stationary RBF (done, but poor performance: 450% error - needs investigation)
- ❌ Matérn kernels (ν = 1/2, 3/2, 5/2)
- ❌ Locally periodic processes
- ❌ 2D spatial fields (elevation, temperature-like)

**Real-World (Validation):**
- ❌ **Motorcycle crash data** (classic non-stationary benchmark, n=133)
- ❌ **Coal mining disasters** (temporal point process, n=191)
- ❌ **Airline passenger data** (periodic + trend, n=144)
- ❌ **Mauna Loa CO2** (long-term trend + seasonality, n=545)
- ❌ **Spatial temperature data** (2D non-stationary field)

### 3. Evaluation Metrics

**Accuracy:**
- Relative Frobenius Error: `||K_pred - K_true|| / ||K_true||` (already implemented)
- Negative Log-Likelihood (NLL) on test set
- Root Mean Squared Error (RMSE) for predictions
- Mean Absolute Error (MAE)

**Efficiency:**
- Training time (seconds)
- Inference time (ms per prediction)
- Memory usage (MB)
- Number of trainable parameters

**Scalability:**
- Performance vs. dataset size (n = 100, 500, 1000, 5000)
- Performance vs. dimensionality (d = 1, 2, 5, 10)

### 4. Deliverables

**Code:**
```
experiments/
├── baselines/
│   ├── rff_baseline.py          # Random Fourier Features
│   ├── standard_gp.py           # Vanilla GP (refactor existing)
│   ├── deep_kernel.py           # (optional)
│   └── sparse_gp.py             # (optional)
├── comparisons/
│   ├── compare_methods.py       # Master comparison script
│   ├── compare_scalability.py   # Scaling experiments
│   └── benchmark_suite.py       # Automated benchmark runner
└── results/
    ├── tables/                  # LaTeX tables
    ├── plots/                   # Comparison plots
    └── raw/                     # Raw results (CSV/JSON)
```

**Paper Assets:**
- **LaTeX comparison table** (Section 4: Experiments)
  - Columns: Method, NLL, RMSE, Time, Memory, Params
  - Rows: One per dataset/method combination
  - Bold best results, underline second-best
- **Comparison plots:**
  - Prediction accuracy vs. training size
  - Training time vs. dataset size
  - Learned spectral densities (F-SDN vs. Remes)
  - Convergence curves

**Documentation:**
- README with instructions to reproduce all results
- Benchmark protocol (hyperparameters, seeds, hardware)
- Analysis of results (where we win/lose and why)

---

## 🔧 Technical Requirements

**Experimental Protocol:**
1. **Train/Test Split:** 80/20 with 5-fold cross-validation
2. **Random Seeds:** Fix to 42, 123, 456, 789, 1011 for reproducibility
3. **Hyperparameter Tuning:** Grid search with validation set
4. **Hardware:** Document GPU/CPU specs for timing comparisons
5. **Early Stopping:** Patience=100 epochs for all methods

**Fair Comparison Guidelines:**
- Use same train/test splits for all methods
- Tune hyperparameters for each method independently
- Report mean ± std over multiple runs (n=5)
- Use best configuration for each method (no intentional handicapping)
- Document all hyperparameters in results table

---

## ✅ Acceptance Criteria

**Minimum Viable Product (MVP):**
- [ ] At least 3 baseline methods implemented
- [ ] At least 5 benchmark datasets (3 synthetic, 2 real-world)
- [ ] Relative Frobenius error computed for all method/dataset pairs
- [ ] One comprehensive LaTeX table for paper
- [ ] Comparison plots showing F-SDN vs. best baseline

**Stretch Goals:**
- [ ] Statistical significance tests (paired t-test, Wilcoxon)
- [ ] Scalability experiments (n vs. time/accuracy)
- [ ] Ablation studies (rank, grid resolution, activation functions)
- [ ] Learned spectral density visualizations
- [ ] Per-dataset analysis with failure case discussion

---

## 📅 Timeline & Priority

**Priority:** 🔴 **CRITICAL** - Blocking NeurIPS submission

**Milestone:** Comparison table by **November 24, 2025** (TODAY!)

**Estimated Effort:** 2-3 days with focused implementation

**Breakdown:**
- **Day 1:** Implement baselines (RFF, refactor Standard GP) - 6h
- **Day 2:** Run benchmark suite on all datasets - 8h
- **Day 3:** Generate tables/plots, write analysis - 4h

---

## 🚧 Known Challenges

1. **RBF Performance:** F-SDN currently shows 450% error on stationary RBF - needs investigation before comparison
2. **Computational Cost:** MC method has O(n²) loops - may need vectorization for large datasets
3. **Hyperparameter Sensitivity:** Need robust grid search to ensure fair comparison
4. **Dataset Availability:** Some real-world datasets may need preprocessing/downloading

---

## 🔗 Related Issues

- Closes #TBD (if there's an existing tracking issue)
- Depends on: None (code is stable)
- Blocks: Final paper submission

---

## 📝 Additional Notes

**Current Experimental Results:**
```
Silverman Non-Stationary (n=100):
  Standard GP:  81.58% error
  F-SDN:        99.03% error  ← Competitive!
  Remes:       173.33% error

Stationary RBF (n=100):
  Standard GP:  46.67% error
  F-SDN:       450.22% error  ← PROBLEM! Needs investigation
  Remes:       212.75% error
```

**Questions to Answer:**
- Why does F-SDN struggle on stationary RBF?
- Is this a fundamental limitation or implementation bug?
- Can we match/beat Remes on non-stationary tasks?
- What's the sweet spot for rank/grid resolution trade-offs?

---

**Assignees:** @AbdullahKaratas, @ArsalanJawaid
**Labels:** `enhancement`, `experiments`, `paper`, `priority:critical`
**Milestone:** NeurIPS 2026 Submission
**Estimated Time:** 2-3 days
